### PR TITLE
Refactor of Rebin

### PR DESCRIPTION
**Pull request by ghost**
**Description of work**
Adds BinningMode option to `Rebin`. This allows you to set the binning mode to use explicitly, to either:
 - "Default": the original behavior, allows for mixed binning types
 - "Linear": will strictly use linear bin width
 - "Logarithmic": will strictly use logarithmic binning
 - "ReverseLogarithmic": will strictly use reverse logarithmic binning
 - "Power": will strictly use inverse power binning, and will fail validation if no power given
 regardless of the sign of the bin width. 

The original behavior is preserved as default, or if set to "Default".

**Purpose of work**

This change adds an optional parameter to Rebin that prevents possible sign errors when working with log-binned data, or other ambiguities.  The goal is to add an explicit flag for the scientist to specify exactly which binning mode they would like to use.  Given their input, the algorithm will force the parameters to behave as the scientist wishes.  This reduces possible user error, as the scientist does not need to keep track of the way flags are set.

This arose as part of work for [EWM1851](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=1851) and work done by @mguthriem in [prototyping the diffraction calibration process](https://github.com/neutrons/prototype_SNAPRed_scripts/blob/main/ewm1126/SNAPRedCalibPrototype.py).

**Summary of work**
New `BinningMode` property added, based on `EnumeratedString`.  Defaullts to `"Default"`, which maintains original behavior.  If a binning mode is specified, then all other input properties are modified to be consistent with the specified mode:
- "Linear", all step sizes in `"Params"` made positive
- "Logarithmic", all step sizes in `"Params"` made negative
- "ReverseLogaritmic", all step sizes in `"Params"` made negative and `"UseReverseLogarithmic"` set to true
- "Power", all step sizes made positive, and fails to validate if power not set

If the power property is set and the binning mode is not `"Default"` or `"Power"`, then the power is reset back to zero.

**Further detail of work**
The static method `Rebin::rebinParamsFromInput`, which is sometimes called by other algorithms, can now also accept an optional string argument for the binning mode.  If specified, it produces rebin params consistent with that specified binning mode.

The static methid  `Rebin::rebinParamsFromInput` is called from the `validateInputs` method.  If not using default binning, then the `validateInputs` method will alter the input binning params to force them to be consistent.

The `InterpolatingRebin` algorithm relied on its parent `Rebin`'s validation methods.  This algo has been given its own validation method, which calls the default behavior of `Rebin::rebinParamsFromInput` and causes a validation error if an exception is thrown.  This does not alter original behavior.

**To test:**
Unit tests have been added to test the new functionallity, specifically you can test:
```ctest -R Rebin```
Running all algorithm tests will take longer, but provide additional validation, as rebin is called implicitly in many, many other algos.

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
As far as I can tell this work is to add extra safety to the algorithm, by making sure that the inputs are correct for the type of binning. So I dont understand why there are changes outside of the validation method. 

@gemmaguest what do you think?

**Comment by rboston628** [comment]
> As far as I can tell this work is to add extra safety to the algorithm, by making sure that the inputs are correct for the type of binning. So I dont understand why there are changes outside of the validation method.
> 
> @gemmaguest what do you think?

You are correct.  But not only to validate the inputs, but also to reinterpret them in light of the specified type of binning.

To explain the motivation, one of our scientists was working on novel calibration method using full-spectrum log-binning.  The calculation required updating the DIFC values, and the formula for that relies on bin width ([see this other PR](https://github.com/mantidproject/mantid/pull/35862)).  To get logarithmic binning, the scientist needed to set his bin with variable to a negative value.  But in his equation using the bin width, he ended up using the negative value without realizing.  The negative sign to bin width has no physical or mathematical meaning, so it's very easy to forget, and took some debugging to find.

This change would allow him to write, for instance:
``` python
dMin = 1
dMax=10
dBin = +1
Rebin(etc. , Params = f"{dMin},{dBin},{dMax}", BinningMode="Logarithmic")
```
and not have to set the bin width parameter to a negative.  In this case, the positive dBin will be reinterpreted as a negative to make it consistent with logarithmic binning.

The reason to not simply return a validation error, is that would force the scientist to set `dBin = -1`.  Now the variable has a negative value that can get lost in a further calculation.

You could also imagine a scientist who inherits some code previously using power binning, like so
``` python
 Rebin( etc.,  Params=f"{dMin},{dBin},{dMax}", Power=0.5)
```
who wants to change to reverse logarithmic binning.  Specifying
``` python
Rebin(etc., Params=f"{dMin},{dBin},{dMax}", Power=0.5, BinningMode="ReverseLogarithmic")
```
will actually reinterpret all of the other properties to make reverse log binning possible; in this case the bin width will be reset to negative and the power set to 0.

**Comment by rboston628** [comment]
In terms of changes outside of the validation method, most of these are cosmetic to enhance the code quality, such as replacing string literal property names with string variable names.  The `rebinParamsFromInput` method was changed to optionally take a binning mode; this method is called by outside algos, and maybe those algos would at some point in the future also like to implement this.  The power being set to zero should probably be part of validation, and not occur inside the `exec` method; that's a change I can make.

The changes to the InterpolatingRebin files are just to give it its own validation method, and an untested use case that came up in the doc tests.

The changes to EnumeratedString are part of [another PR](https://github.com/mantidproject/mantid/pull/35904) that are only included to allow this to run.  They will disappear soon.  

The change to the SimpleAPITest is to fix a docstring test, that tests the Rebin docs.  Since a property was added, the docstring changed, so this test needed to change.

**Comment by rboston628** [comment]
With the exception of cosmetic improvements to code quality, the `exec` method of Rebin is no unchanged by this PR.  All changes have been moved to the helper method to get the bin params, and the vadliation method.

File: Framework/Algorithms/test/RebinTest.h [hunk 1]
```diff
@@ -118,6 +118,59 @@ class RebinTest : public CxxTest::TestSuite {
...
+    rebin.initialize();
+    rebin.setPropertyValue("InputWorkspace", "test_in1D");
+    rebin.setPropertyValue("OutputWorkspace", "test_out");
+    rebin.setPropertyValue("Params", "2.0");
+    TS_ASSERT_THROWS_NOTHING(rebin.execute());
+    TS_ASSERT(rebin.isExecuted());
+
+    // ensure that the properties are not changed in default mode
+    std::vector<double> rbParams = rebin.getProperty("Params");
+    TS_ASSERT_EQUALS(rbParams.size(), 1);
```
Comment (jfkcooper) [inline Framework/Algorithms/test/RebinTest.h:136]: isn't this testing that rbParams is of size 1, not 3?

File: Framework/Algorithms/test/RebinTest.h [hunk 2]
```diff
@@ -118,6 +118,59 @@ class RebinTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_EQUALS(2.0, rbParams[0]);
+
+    AnalysisDataService::Instance().remove("test_in1D");
+    AnalysisDataService::Instance().remove("test_out");
+  }
+
+  void test_make_three_params_modeset() {
+    // test that if only a binwidth is given, will create an xmin, xmax from xstep
```
Comment (jfkcooper) [inline Framework/Algorithms/test/RebinTest.h:144]: this has the same comment as the function above, but I don't think does the same thing?

File: Framework/Algorithms/test/RebinTest.h [hunk 3]
```diff
@@ -146,7 +199,139 @@ class RebinTest : public CxxTest::TestSuite {
...
+        AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>("test_out_default");
+    auto &outX_default = rebindata_default->x(0);
+
+    // now run again but with positive step, using logarithmic mode
+    Rebin rebin_again;
+    rebin_again.initialize();
+    rebin_again.setPropertyValue("InputWorkspace", "test_in1D");
+    rebin_again.setPropertyValue("OutputWorkspace", "test_out_again");
+    rebin_again.setPropertyValue("Params", "1.0,1.0,1000.0");
+    rebin_again.setPropertyValue("Power", "0.5");
```
Comment (jfkcooper) [inline Framework/Algorithms/test/RebinTest.h:232]: should the testing of the power being set to zero have its own test?

File: Framework/Algorithms/test/RebinTest.h [hunk 4]
```diff
@@ -146,7 +199,139 @@ class RebinTest : public CxxTest::TestSuite {
...
+    auto &outX_default = rebindata_default->x(0);
+
+    // now run again but with positive step, using logarithmic mode
+    Rebin rebin_again;
+    rebin_again.initialize();
+    rebin_again.setPropertyValue("InputWorkspace", "test_in1D");
+    rebin_again.setPropertyValue("OutputWorkspace", "test_out_again");
+    rebin_again.setProperty("Params", "1.0,1.0,1000.0");
+    rebin_again.setProperty("UseReverseLogarithmic", false);
+    rebin_again.setProperty("Power", "0.5");
```
Comment (jfkcooper) [inline Framework/Algorithms/test/RebinTest.h:286]: Is the power here used at all?

**Review by jfkcooper — COMMENTED** [review]
A comment that I am neither a mantid, nor a CPP developer, so please take my comments as such. (I'm also only commenting on the test code due to this)
Since rebin is such an integral part of many parts of both Mantid as well as user scripts I would try to make sure that all edge cases are covered in the tests, and I believe there are a few cases which are not being covered.
As I see it ,there are three different use cases for rebin: linear, power, and logarithmic. I think that all three should have their behaviour compared with and without the mode_set to ensure that it is indeed the same, this should also be done on partial as well as full datasets - i.e. either just providing the single value for Params, or a set of values which ensures that the edge points are definitely not on bin edges.
I would also add a couple more tests for checking that the error behaviour is maintained: e.g. what happens if Params has 4 values, or none, or a single value but it is zero.
One of the tests seems to be testing both that "Power" gets set back to zero, as well as checking the values are the same for logarithmic binning with and without mode_set, however this confuses the output a bit since the old Rebin behaviour was that having a non-zero "Power" with logarithmic binning would throw an error. A separate test to check that logarithmic binning with Power unset, gives the same value as when it is set should be added for this case.
There are also the cases for the various options of Rebin, whose behaviour should be checked that it is maintained (PreserveEvents, FullBinsOnly,IgnoreBinErrors, UseReverseLogarithm). Of these, there is a test which touches UseReverseLogarithm, but I don't think compares its behaviour in the cases of it being True with and without the mode_set.

Anyway, lots of comments and suggestions, I'm happy to discuss any if it isn't clear what I mean by them.

File: Framework/Algorithms/test/RebinTest.h (hunk 1 above)
Comment (rboston628) [reply Framework/Algorithms/test/RebinTest.h:136]: There are two things tested here.  The first set of asserts ensures that the default algorithm will create a full set of rebin params if only given a step size.  This is verified by the algorithm being able to execute without errors.  This was previously untested behavior.

The second set of asserts verifies that the "Params" property was not changed.  This is the original behavior of this algo, and needs to remain preserved.

File: Framework/Algorithms/test/RebinTest.h (hunk 4 above)
Comment (rboston628) [reply Framework/Algorithms/test/RebinTest.h:286]: It is not used by the algorithm.  It is being set to verify that if the binning mode is reverse logarithmic, the algo will not be confused by nor use the power property.

File: Framework/Algorithms/test/RebinTest.h (hunk 3 above)
Comment (rboston628) [reply Framework/Algorithms/test/RebinTest.h:232]: I thought that it did.  If it does not, then you are correct, that should be a specific test.

File: Framework/Algorithms/test/RebinTest.h (hunk 1 above)
Comment (jfkcooper) [reply Framework/Algorithms/test/RebinTest.h:136]: I think I am just a bit confused (if I'm reading it correctly) by the differing behaviour between this and e.g. L165, where Params is still being given a single value, but constructs the xmin and xmax from the data, sorry I should have been clearer with the comment

**Comment by rboston628** [comment]
@jfkcooper for log, reverse log, and power, there is a test that runs the algo under the original settings, then runs it with the modeset, and ensures all bin edges are the same in the two output workspaces.  I can agree that there is too much going on in some of the unit tests, so I will make steps to simplify them, as well as add more testing generally.

**Comment by rboston628** [comment]
@jfkcooper

These are the tests I think you're asking for:
 - compare default to mode set, w/ single rebin param
 - compare default to mode set, w/ params so edge points not on bin edges
 - test failures of Params under some invalid conditions
 - check log binning gives same output, with and without "Power" set
 - check "UseReverseLogarithm" made true whenever BinningMode=ReverseLogarithmic
 
Because all changes of behavior occur inside the validateInputs() method, the number of tests can be reduced some.

There are two input validation steps in mantid.  One is for property validators, which are executed when the property is set.  The other is the validator method, which is run just before execution of the algorithm.
 
I made changes to the rebinning parameters validator to include a single zero, and added the tests you requested.  These tests fail when the property is set, so do not require testing with the bin mode set, nor with executing the algorithm.

Since only the validator method changes properties, so it should suffice to check:
 1. after the validator, properties are changed to be consistent with desired behavior
 2. executing the algorithm doesn't change any of the properties
 3. executing the algorithm with the correct properties produces expected results
 
I added in tests especially for Steps 1 and 2.  Step 3 is done by all of the tests of Rebin that already exist.  I also have tests that runs the algo in default mode with the properties set for a binning mode, then runs it with the binning mode set, and verifies the same output is produced.

I added a test of the default behavior, that mixing log and power binning would produce an error.  For peace of mind, I added a three-way test for all non-Power binning modes, that when run with our without the power set they produce the same results as binning under default mode.  This isn't strictly needed, but is useful to have.

All told, I think the unit tests have been significantly improved following your comment, with further improvements to some other part of code touched by rebin.

File: Framework/Algorithms/test/RebinTest.h (hunk 1 above)
Comment (rboston628) [reply Framework/Algorithms/test/RebinTest.h]: This part is marked outdated, but the tests still exist in L53 and L970.  In Default mode, the `exec` method should fill in the endpoints of the rebin parameters using the workspace endpoints and use them internally, but not change the property.  In every other binning mode, the `validateInputs` method should fill in the endpoints and change the property before `exec` runs.  I think it's more clean in the refactor since your comments.

**Comment by jfkcooper** [comment]
Thanks for implementing these @rboston628! I'm pretty sure these cover all cases that I can think of.

**Comment by rboston628** [comment]
The changes to the rebin params validator were moved to [PR 35934](https://github.com/mantidproject/mantid/pull/35934)

**Review by jfkcooper — APPROVED** [review]
Changes to testing of rebin both covers my concerns and improves the code. Many thanks

File: Framework/Algorithms/src/InterpolatingRebin.cpp [hunk 5]
```diff
@@ -54,8 +54,26 @@ void InterpolatingRebin::init() {
...
+ */
+/// Validate that the input properties are sane.
+std::map<std::string, std::string> InterpolatingRebin::validateInputs() {
+  std::map<std::string, std::string> helpMessages;
+
+  // as part of validation, force the binwidth to be compatible with set bin mode
+  MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  std::vector<double> rbParams = getProperty("Params");
+  try {
+    std::vector<double> validParams = Rebin::rebinParamsFromInput(rbParams, *inputWS, g_log);
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/InterpolatingRebin.cpp:69]: To follow from what Rebin does in it's `validateInputs()`, I think think you need to use the `validParams` to set the `Param` property like `Rebin` does. i.e.
```c
      setProperty("Params", validParams);
```

File: Framework/Algorithms/test/InterpolatingRebinTest.h [hunk 6]
```diff
@@ -96,6 +96,22 @@ class InterpolatingRebinTest : public CxxTest::TestSuite {
...
+    test_in1D->setDistribution(true);
+    AnalysisDataService::Instance().add("InterpolatingRebinTest_indist", test_in1D);
+
+    InterpolatingRebin rebin;
+    rebin.initialize();
+    rebin.setPropertyValue("InputWorkspace", "InterpolatingRebinTest_indist");
+    rebin.setPropertyValue("OutputWorkspace", "InterpolatingRebinTest_outdist");
+    rebin.setPropertyValue("Params", "1.5");
+    TS_ASSERT_THROWS_NOTHING(rebin.execute())
+    TS_ASSERT(rebin.isExecuted())
```
Comment (peterfpeterson) [inline Framework/Algorithms/test/InterpolatingRebinTest.h:112]: This test should have a a check for what the `Params` is coming back from this

File: Framework/Algorithms/test/RebinTest.h [hunk 7]
```diff
@@ -48,6 +48,70 @@ class RebinTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_THROWS_ANYTHING(rebin.setPropertyValue("Params", "-1.0,-1.0,1000.0"));
+    TS_ASSERT(!rebin.isExecuted());
+  }
+
+  void test_failure_bad_rebin_params() {
+    // ensure that rebinning will fail if log binning from negative to positive
+    // this fails fast, due to the rebin param property validator
+    Rebin rebin;
+    rebin.initialize();
+    TS_ASSERT_THROWS_ANYTHING(rebin.setPropertyValue("Params", "-1.0,-1.0,10.0"));
```
Comment (peterfpeterson) [inline Framework/Algorithms/test/RebinTest.h:91]: Can this be done with initializer lists? 
```
    TS_ASSERT_THROWS_ANYTHING(rebin.setProperty("Params", {-1.0,-1.0,10.0}));
```
or
```
    TS_ASSERT_THROWS_ANYTHING(rebin.setProperty("Params", std::vector<double>{-1.0,-1.0,10.0}));
```
?

File: Framework/Algorithms/test/RebinTest.h [hunk 8]
```diff
@@ -748,6 +812,537 @@ class RebinTest : public CxxTest::TestSuite {
...
+      for (char only_step = 0; only_step < 2; only_step++) {
+        do_test_changes_rebin_params(binModes[i], steps[i], goodSteps[i], only_step);
+      }
+    }
+  }
+
+  void do_test_changes_power(std::string binMode, double step) {
+    // check that algo run in binning  mode will change power
+    std::stringstream params;
+    params << "1.0," << step << ",100.0";
```
Comment (peterfpeterson) [inline Framework/Algorithms/test/RebinTest.h:1030]: This could be done using a `std::vector<double>`

File: Framework/Algorithms/test/RebinTest.h [hunk 9]
```diff
@@ -748,6 +812,537 @@ class RebinTest : public CxxTest::TestSuite {
...
+
+    Workspace2D_sptr test_in1D = Create1DWorkspace(50);
+    AnalysisDataService::Instance().add("test_in1D", test_in1D);
+
+    Rebin rebin;
+    rebin.initialize();
+    rebin.setPropertyValue("InputWorkspace", "test_in1D");
```
Comment (peterfpeterson) [inline Framework/Algorithms/test/RebinTest.h:1037]: Debugging is a little easier if you set the workspace names to be similar to the test name. Also, it can be good to define the names as string constants at the top of the test, e.g.
```c
const std::string IN_WKSP("changes_power_in");
const std::string OUT_WKSP("changes_power_out");
```

File: Framework/Algorithms/test/InterpolatingRebinTest.h (hunk 6 above)
Comment (peterfpeterson) [reply Framework/Algorithms/test/InterpolatingRebinTest.h:112]: You were right. Please add a tiny check.
Above this part
```c
constexpr double WIDTH{1.5}
...
rebin.setProperty("Params", WIDTH);
```

```c
    std::vector<double> params = rebin.getProperty("Params");
    TS_ASSERT_EQUALS(params.size(), 1);
    TS_ASSERT_EQUALS(params.front(), WIDTH);
```

File: Framework/Algorithms/test/RebinTest.h (hunk 7 above)
Comment (rboston628) [reply Framework/Algorithms/test/RebinTest.h:91]: The second one works

File: Framework/Algorithms/test/RebinTest.h (hunk 8 above)
Comment (rboston628) [reply Framework/Algorithms/test/RebinTest.h:1030]: much cleaner, in hingsight

**Review by peterfpeterson — APPROVED** [review]
I like the new option. It makes the binning mode clearer than having special values in the parameters.
